File: business-sales-backend/src/models/SaleRecord.ts
import mongoose, { Schema, Document, type InferSchemaType } from 'mongoose';
import { MenuCategory, ItemSize } from '../types';

// Sale Item Schema - for multi-item sales
const SaleItemSchema = new Schema({
  menuItemId: {
    type: Schema.Types.ObjectId,
    required: true,
    ref: 'MenuItems'
  },
  itemName: {
    type: String,
    required: true,
    trim: true
  },
  categoryId: {
    type: Schema.Types.ObjectId,
    required: true,
    ref: 'Categories'
  },
  categoryName: {
    type: String,
    required: true,
    trim: true
  },
  size: {
    type: String,
    required: true,
    enum: Object.values(ItemSize)
  },
  unitPrice: {
    type: Number,
    required: true,
    min: 0
  },
  quantity: {
    type: Number,
    required: true,
    min: 1
  },
  subtotal: {
    type: Number,
    required: true,
    min: 0
  }
}, { _id: false });

const SaleRecordSchema = new Schema({
  userId: {
    type: Schema.Types.ObjectId,
    required: true,
    ref: 'Users',
    index: true
  },
  // Single-item fields (for backward compatibility and quick sales)
  menuItemId: {
    type: Schema.Types.ObjectId,
    ref: 'MenuItems'
  },
  itemName: {
    type: String,
    trim: true
  },
  categoryId: {
    type: Schema.Types.ObjectId,
    ref: 'Categories'
  },
  categoryName: {
    type: String,
    trim: true,
    maxlength: 50
  },
  size: {
    type: String,
    enum: Object.values(ItemSize)
  },
  unitPrice: {
    type: Number,
    min: 0
  },
  quantity: {
    type: Number,
    min: 1
  },
  // Multi-item fields (for POS/billing transactions)
  items: {
    type: [SaleItemSchema],
    validate: {
      validator: function(this: any, items: any[]) {
        // If items array exists, it must have at least one item
        // If it doesn't exist, single-item fields must be present
        if (items && items.length > 0) return true;
        if (!items || items.length === 0) {
          return this.menuItemId != null;
        }
        return false;
      },
      message: 'Sale must have either items array or single item fields'
    }
  },
  // Shared fields
  totalAmount: {
    type: Number,
    required: true,
    min: 0
  },
  taxAmount: {
    type: Number,
    default: 0,
    min: 0,
    required: false // Make it explicitly optional
  },
  grandTotal: {
    type: Number,
    min: 0,
    required: false // Make it optional for backward compatibility
  },
  timestamp: {
    type: Date,
    default: Date.now,
    required: true
  },
  notes: {
    type: String,
    trim: true,
    maxlength: 500
  },
  // Invoice tracking fields
  invoiceNumber: {
    type: String,
    trim: true,
    index: true,
    unique: true,
    sparse: true // Allows null values but enforces uniqueness when present
  },
  invoiceGenerated: {
    type: Boolean,
    default: false
  },
  invoiceGeneratedAt: {
    type: Date
  },
  // Payment information
  paymentMethod: {
    type: String,
    enum: ['cash', 'card', 'upi', 'other'],
    default: 'cash'
  },
  paymentStatus: {
    type: String,
    enum: ['pending', 'paid', 'partially_paid', 'refunded'],
    default: 'paid'
  },
  // Type indicator
  saleType: {
    type: String,
    enum: ['single', 'multi'],
    required: true,
    default: function(this: any) {
      return this.items && this.items.length > 0 ? 'multi' : 'single';
    }
  }
},{
  timestamps: true
});

// Pre-save hook to ensure grandTotal is set for old records
SaleRecordSchema.pre('save', function(next) {
  // If grandTotal is not set, calculate it
  if (this.grandTotal == null || this.grandTotal === undefined) {
    const taxAmount = this.taxAmount || 0;
    this.grandTotal = this.totalAmount + taxAmount;
  }
  next();
});

// Post-find hooks to ensure grandTotal is present for old records when reading
const ensureGrandTotal = function(doc: any) {
  if (doc && (doc.grandTotal == null || doc.grandTotal === undefined)) {
    const taxAmount = doc.taxAmount || 0;
    doc.grandTotal = doc.totalAmount + taxAmount;
  }
};

SaleRecordSchema.post('find', function(docs) {
  if (Array.isArray(docs)) {
    docs.forEach(ensureGrandTotal);
  }
});

SaleRecordSchema.post('findOne', function(doc) {
  ensureGrandTotal(doc);
});

SaleRecordSchema.post('findOneAndUpdate', function(doc) {
  ensureGrandTotal(doc);
});

// Generate the type from the schema
type SaleRecordSchemaType = InferSchemaType<typeof SaleRecordSchema>;

// Create the interface extending Document and the schema type
interface ISaleRecord extends SaleRecordSchemaType, Document { }


// Indexes for analytics and queries
SaleRecordSchema.index({ timestamp: -1 }); // Latest sales first
SaleRecordSchema.index({ menuItemId: 1 });
SaleRecordSchema.index({ categoryId: 1 }); // New index for dynamic categories
SaleRecordSchema.index({ userId: 1, categoryId: 1 }); // User's category sales
SaleRecordSchema.index({ timestamp: 1, categoryId: 1 }); // Compound index for analytics
SaleRecordSchema.index({ categoryName: 1 }); // Text-based category queries
SaleRecordSchema.index({ saleType: 1 }); // Filter by sale type
SaleRecordSchema.index({ userId: 1, saleType: 1 }); // User's sales by type
SaleRecordSchema.index({ userId: 1, invoiceGenerated: 1 }); // User's invoices

// Legacy indexes (to be removed after migration)
SaleRecordSchema.index({ category: 1 });
SaleRecordSchema.index({ timestamp: 1, category: 1 });

// Create date-based indexes for faster analytics
SaleRecordSchema.index({ 
  timestamp: 1,
  categoryId: 1,
  totalAmount: 1 
});

export default mongoose.model<ISaleRecord>('SaleRecords', SaleRecordSchema); 

File: flutter_app/lib/models/sale_record.dart
import 'package:json_annotation/json_annotation.dart';
import 'menu_item.dart';

part 'sale_record.g.dart';

// Sale Item for multi-item sales
@JsonSerializable()
class SaleItem {
  final String menuItemId;
  final String itemName;
  final String categoryId;
  final String categoryName;
  final ItemSize size;
  final double unitPrice;
  final int quantity;
  final double subtotal;

  const SaleItem({
    required this.menuItemId,
    required this.itemName,
    required this.categoryId,
    required this.categoryName,
    required this.size,
    required this.unitPrice,
    required this.quantity,
    required this.subtotal,
  });

  factory SaleItem.fromJson(Map<String, dynamic> json) =>
      _$SaleItemFromJson(json);
  Map<String, dynamic> toJson() => _$SaleItemToJson(this);
}

@JsonSerializable()
class SaleRecord {
  @JsonKey(name: '_id')
  final String id;

  // Single-item fields (for backward compatibility and quick sales)
  final String? menuItemId;
  final String? itemName;
  final String? categoryId; // Reference to Category._id (New Dynamic System)
  final String? categoryName; // Category name for faster querying
  final MenuCategory? category; // Legacy field for backwards compatibility
  final ItemSize? size;
  final double? unitPrice;
  final int? quantity;

  // Multi-item fields
  final List<SaleItem>? items;
  final String saleType; // 'single' or 'multi'

  // Shared fields
  final double totalAmount;
  final double? taxAmount;
  final double grandTotal;
  final DateTime timestamp;
  final String? notes;
  final String? userId; // Backend userId field
  final DateTime? createdAt;
  final DateTime? updatedAt;

  // Invoice fields
  final String? invoiceNumber;
  final bool? invoiceGenerated;
  final DateTime? invoiceGeneratedAt;

  // Payment fields
  final String? paymentMethod;
  final String? paymentStatus;

  const SaleRecord({
    required this.id,
    this.menuItemId,
    this.itemName,
    this.categoryId,
    this.categoryName,
    this.category,
    this.size,
    this.unitPrice,
    this.quantity,
    this.items,
    this.saleType = 'single',
    required this.totalAmount,
    this.taxAmount,
    required this.grandTotal,
    required this.timestamp,
    this.notes,
    this.userId,
    this.createdAt,
    this.updatedAt,
    this.invoiceNumber,
    this.invoiceGenerated,
    this.invoiceGeneratedAt,
    this.paymentMethod,
    this.paymentStatus,
  });

  factory SaleRecord.fromJson(Map<String, dynamic> json) {
    // Handle the backend response format
    Map<String, dynamic> processedJson = Map<String, dynamic>.from(json);

    // Convert MongoDB _id to id
    if (processedJson.containsKey('_id') && processedJson['_id'] is Map) {
      processedJson['_id'] = processedJson['_id']['\$oid'];
    }

    // Convert menuItemId ObjectId to string
    if (processedJson.containsKey('menuItemId') &&
        processedJson['menuItemId'] is Map) {
      processedJson['menuItemId'] = processedJson['menuItemId']['\$oid'];
    }

    // Convert categoryId ObjectId to string
    if (processedJson.containsKey('categoryId') &&
        processedJson['categoryId'] is Map) {
      processedJson['categoryId'] = processedJson['categoryId']['\$oid'];
    }

    // Convert userId ObjectId to string
    if (processedJson.containsKey('userId') && processedJson['userId'] is Map) {
      processedJson['userId'] = processedJson['userId']['\$oid'];
    }

    // Convert date strings to DateTime
    if (processedJson.containsKey('timestamp')) {
      if (processedJson['timestamp'] is Map &&
          processedJson['timestamp'].containsKey('\$date')) {
        processedJson['timestamp'] = processedJson['timestamp']['\$date'];
      }
    }

    if (processedJson.containsKey('createdAt')) {
      if (processedJson['createdAt'] is Map &&
          processedJson['createdAt'].containsKey('\$date')) {
        processedJson['createdAt'] = processedJson['createdAt']['\$date'];
      }
    }

    if (processedJson.containsKey('updatedAt')) {
      if (processedJson['updatedAt'] is Map &&
          processedJson['updatedAt'].containsKey('\$date')) {
        processedJson['updatedAt'] = processedJson['updatedAt']['\$date'];
      }
    }

    // Handle missing fields for backward compatibility
    // Set default taxAmount if missing
    if (!processedJson.containsKey('taxAmount') ||
        processedJson['taxAmount'] == null) {
      processedJson['taxAmount'] = 0.0;
    }

    // Calculate grandTotal if missing (for old records)
    if (!processedJson.containsKey('grandTotal') ||
        processedJson['grandTotal'] == null) {
      final totalAmount = processedJson['totalAmount'] ?? 0.0;
      final taxAmount = processedJson['taxAmount'] ?? 0.0;
      processedJson['grandTotal'] = totalAmount + taxAmount;
    }

    // Set default saleType if missing
    if (!processedJson.containsKey('saleType') ||
        processedJson['saleType'] == null) {
      // Determine saleType based on presence of items array
      if (processedJson.containsKey('items') &&
          processedJson['items'] is List &&
          (processedJson['items'] as List).isNotEmpty) {
        processedJson['saleType'] = 'multi';
      } else {
        processedJson['saleType'] = 'single';
      }
    }

    return _$SaleRecordFromJson(processedJson);
  }

  Map<String, dynamic> toJson() => _$SaleRecordToJson(this);

  // Factory constructor to create a single-item sale record from menu item for API submission
  factory SaleRecord.fromMenuItem({
    required String id,
    required MenuItem menuItem,
    required String categoryName, // Must be provided from Category object
    required ItemSize size,
    required int quantity,
    required DateTime timestamp,
    String? notes,
    String? userId,
  }) {
    final unitPrice = menuItem.getPriceBySize(size);
    final totalAmount = unitPrice * quantity;

    return SaleRecord(
      id: id,
      menuItemId: menuItem.id,
      itemName: menuItem.name,
      categoryId: menuItem.categoryId,
      categoryName: categoryName,
      size: size,
      unitPrice: unitPrice,
      quantity: quantity,
      saleType: 'single',
      totalAmount: totalAmount,
      grandTotal: totalAmount,
      timestamp: timestamp,
      notes: notes,
      userId: userId,
    );
  }

  // Factory constructor to create a multi-item sale record
  factory SaleRecord.fromItems({
    required String id,
    required List<SaleItem> items,
    required String userId,
    required DateTime timestamp,
    String? notes,
    double taxAmount = 0,
    String paymentMethod = 'cash',
    String paymentStatus = 'paid',
  }) {
    final totalAmount = items.fold<double>(
      0,
      (sum, item) => sum + item.subtotal,
    );
    final grandTotal = totalAmount + taxAmount;

    return SaleRecord(
      id: id,
      items: items,
      saleType: 'multi',
      totalAmount: totalAmount,
      taxAmount: taxAmount,
      grandTotal: grandTotal,
      timestamp: timestamp,
      notes: notes,
      userId: userId,
      paymentMethod: paymentMethod,
      paymentStatus: paymentStatus,
    );
  }

  // Helper to check if this is a multi-item sale
  bool get isMultiItem =>
      saleType == 'multi' && items != null && items!.isNotEmpty;

  // Helper to check if this is a single-item sale
  bool get isSingleItem => saleType == 'single' && menuItemId != null;

  // Helper methods for date filtering
  bool isToday() {
    final now = DateTime.now();
    return timestamp.year == now.year &&
        timestamp.month == now.month &&
        timestamp.day == now.day;
  }

  bool isThisWeek() {
    final now = DateTime.now();
    final weekStart = now.subtract(Duration(days: now.weekday - 1));
    return timestamp.isAfter(weekStart.subtract(const Duration(days: 1)));
  }

  bool isThisMonth() {
    final now = DateTime.now();
    return timestamp.year == now.year && timestamp.month == now.month;
  }

  bool isThisYear() {
    final now = DateTime.now();
    return timestamp.year == now.year;
  }
}
